src/cdk/a11y/key-manager/typeahead.ts

```diff
@@ -0,0 +1,118 @@
+import {Subject, Subscription} from 'rxjs';
+import {A, Z, ZERO, NINE} from '@angular/cdk/keycodes';
+import {debounceTime, filter, map, take, tap} from 'rxjs/operators';
+
+const DEFAULT_TYPEAHEAD_DEBOUNCE_INTERVAL_MS = 200;
+
+interface TypeaheadItem {
+  getLabel?(): string;
+}
+
+interface TypeaheadConfig<T> {
+  debounceInterval?: number;
+  skipPredicate?: (item: T) => boolean | undefined;
+}
+
+export class Typeahead<T extends TypeaheadItem> {
+  private readonly _letterKeyStream = new Subject<string>();
+  private _items: T[] = [];
+  private _selectedItemIndex = -1;
+
+  /** Buffer for the letters that the user has pressed */
+  private _pressedLetters: string[] = [];
+
+  private _skipPredicateFn?: (item: T) => boolean | undefined;
+
+  private readonly _selectedItem = new Subject<T>();
+  readonly selectedItem = this._selectedItem.asObservable();
+
+  constructor(initialItems: T[], config?: TypeaheadConfig<T>) {
+    const typeAheadInterval =
+      typeof config?.debounceInterval === 'number'
+        ? config.debounceInterval
+        : DEFAULT_TYPEAHEAD_DEBOUNCE_INTERVAL_MS;
+
+    if (config?.skipPredicate) {
+      this._skipPredicateFn = config.skipPredicate;
+    }
+
+    this.setItems(initialItems);
+    this._setupKeyHandler(typeAheadInterval);
+  }
+
+  destroy() {
+    this._pressedLetters = [];
+    this._letterKeyStream.complete();
+    this._selectedItem.complete();
+  }
+
+  setCurrentSelectedItemIndex(index: number) {
+    this._selectedItemIndex = index;
+  }
+
+  setItems(items: T[]) {
+    if (
+      (typeof ngDevMode === 'undefined' || ngDevMode) &&
+      items.length &&
+      items.some(item => typeof item.getLabel !== 'function')
+    ) {
+      throw new Error('KeyManager items in typeahead mode must implement the `getLabel` method.');
+    }
+
+    this._items = items;
+  }
+
+  handleKey(event: KeyboardEvent): void {
+    const keyCode = event.keyCode;
+
+    // Attempt to use the `event.key` which also maps it to the user's keyboard language,
+    // otherwise fall back to resolving alphanumeric characters via the keyCode.
+    if (event.key && event.key.length === 1) {
+      this._letterKeyStream.next(event.key.toLocaleUpperCase());
+    } else if ((keyCode >= A && keyCode <= Z) || (keyCode >= ZERO && keyCode <= NINE)) {
+      this._letterKeyStream.next(String.fromCharCode(keyCode));
+    }
+  }
+
+  /** Gets whether the user is currently typing into the manager using the typeahead feature. */
+  isTyping(): boolean {
+    return this._pressedLetters.length > 0;
+  }
+
+  // TODO: find a better name?
+  reset(): void {
```

Comment:
I think reset is an appropriate name. What this does is allows the end user to restart typing by reseting the state of type letters to the initial state.